EIPS/eip-2585.md

```diff
@@ -0,0 +1,443 @@
+            require(checkAndUpdateNonce(message.from, message.nonce), "NONCE_INVALID");
+        } else {
+            require(ReplayProtection(message.replayProtection).checkAndUpdateNonce(message.from, message.nonce), "NONCE_INVALID");
+        }
+
+        _call(message.from, message.to, msg.value, message.data);
+    }
+
+
+    // /////////////////////////////////// BATCH CALL /////////////////////////////////////
+
+    struct Call {
+        address to;
+        bytes data;
+        uint256 value;
+    }
+
+    /// @notice batcher function that can be called as part of a meta transaction (allowing to batch call atomically)
+    /// @param calls list of call data and destination
+    function batch(Call[] memory calls) public payable { // external with ABIEncoderV2 Struct is not supported in solidity < 0.6.4
+        require(msg.sender == address(this), "FORWARDER_ONLY");
+        address signer;
+        bytes memory data = msg.data;
+        uint256 length = msg.data.length;
+        assembly { signer := and(mload(sub(add(data, length), 0x00)), 0xffffffffffffffffffffffffffffffffffffffff) }
+        for(uint256 i = 0; i < calls.length; i++) {
+            _call(signer, calls[i].to, calls[i].value, calls[i].data);
+        }
+    }
+
+    // /////////////////////////////////// REPLAY PROTECTION /////////////////////////////////////
+
+    mapping(address => mapping(uint128 => uint128)) _batches;
+
+    /// @notice implement a default nonce stategy
+    /// @param signer address to check and update nonce for
+    /// @param nonce value of nonce sent as part of the forward call
+    function checkAndUpdateNonce(address signer, bytes memory nonce) public override returns (bool) {
+        // TODO? default nonce strategy could be different (maybe the most versatile : batchId + Nonce)
+        uint256 value = abi.decode(nonce, (uint256));
+        uint128 batchId = uint128(value / 2**128);
+        uint128 batchNonce = uint128(value % 2**128);
+
+        uint128 currentNonce = _batches[signer][batchId];
+        if (batchNonce == currentNonce) {
+            _batches[signer][batchId] = currentNonce + 1;
+            return true;
+        }
+        return false;
+    }
+
+    function getNonce(address signer, uint128 batchId) external view returns (uint128) {
+        return _batches[signer][batchId];
+    }
+
+
+    // ///////////////////////////////// INTERNAL ////////////////////////////////////////////
+
+    function _call(
+        address from,
+        address to,
+        uint256 value,
+        bytes memory data
+    ) internal {
+        (bool success,) = to.call.value(value)(abi.encodePacked(data, from));
+        if (!success) {
+            assembly {
+                let returnDataSize := returndatasize()
+                returndatacopy(0, 0, returnDataSize)
+                revert(0, returnDataSize)
+            }
+        }
+    }
+
+    function _checkSigner(
```

Comment:
I've been doing that without a sigatureType using try catch. You may want to consider something like:

	if (_isContract(message.from))
	{
		try IERC1271(message.from).isValidSignature(dataToHash, _signature) returns (bytes4 value)
		{
			require(value == IERC1271(0).isValidSignature.selector, "SIGNATURE_1271_INVALID");
		}
		catch (bytes memory /*lowLevelData*/) {}
		try IERC1654(message.from).isValidSignature(keccak256(dataToHash), _signature) returns (bytes4 value)
		{
			require(value == IERC1654(0).isValidSignature.selector, "SIGNATURE_1654_INVALID");
		}
		catch (bytes memory /*lowLevelData*/) {}
		revert("NO_SUPPORTED_CONTRACT_SIGNATURE");
	}
	else
	{
		require(message.from == SigUtil.recover(keccak256(dataToHash), signature, "SIGNATURE_WRONG_SIGNER");
	}

Reply:
I thought of doing it that way but preferred being explicit, but I am open to change.

But actually thinking of it, the signature type should be part of the data to be signed. 
The reasoning is mentioned here : https://github.com/ethereum/EIPs/issues/1271#issuecomment-531472557

Basically if a wallet support both EIP-1654 and EIP-1271, then the signer should be protected from someone using the EIP-1654 version instead of the EIP-1271 one (which can according to the  EIP-1271 standard have more strict rules like time based spending restriction, etc... (even though I am not sure how it practise it would be achieved). 
So if the signature type is not part of the data, the relayer would be able to bypass the stricter 1271 one. In your implementation this would actually be hard to do since the call to EIP-1271 is done before the EIP-1654 but not impossible. Indeed, because of the EIP-150 1/64 rules there could be condition where a gas heavy 1271 call would fails due to not enough gas passed (maybe in purpose by the relayer), resulting in the call to EIP-1654 to succeed. Quite contrived I agree, but still theoretically possible

As for signature type, we could use EIP-2126 : https://ethereum-magicians.org/t/erc-2126-signature-type-recognition/3392 ?

Reply:
I having a hard time imagining a wallet implementing both erc1271 and erc1654 and not enforcing the same security level on both front. But you are right, almost no wallet implement either, so we are very far from seing both implementation waltking on each other's foot.

My approach was:
- if strict ERC1271 is available, we use that, and we are good (return, don't do further checks)
- if less strict ERC1654 is available, we use that, and we are good (return, don't do further checks)
- oherwize we don't know how to process the signature, so we revert

Also, I have a VERY string disagreament with ERC2126, which was never adressed despite my many comments on the ethmagician forum. ERC2126 ask you to decide between `0x4` (Wallet) and `0x2` (ERC712). So if you want to have a erc712 signature recognized by a wallet through either erc1271 or erc1654, you have no way to express that.
Knowing how to hash the data is an app issue, knowing how to check the signature of this hashed data is something different. ERC2126 is considering both as the same issue, and there is no way I'm going to support that until there is a clear distinction made.